capi: support Flatcar OVA images

johananl opened the pull request:
What this PR does / why we need it:
This PR adds Flatcar as one of the distros in the VMware OVA image builder.

Which issue(s) this PR fixes (optional, in fixes #<issue number>(, fixes #<issue_number>, ...) format, will close the issue(s) when PR gets merged): None

**Additional context**

TODOs:

- [x] Add Flatcar as a CI [target](https://github.com/kubernetes-sigs/image-builder/blob/142e815b24593287025b6a53ff3f2c6625d7cfac/images/capi/scripts/ci-ova.sh#L24)?
- [ ] I'm having a hard time testing the `build-node-ova-local-flatcar` target because I can't run VMware Workstation on my machine. Should this be a blocker?
- [x] Fix GOSS errors (the build passes though).
- [x] Ensure Flatcar builds OK for all supported platforms.
- [x] Add Flatcar to https://image-builder.sigs.k8s.io/capi/providers/vsphere.html#configuration.
- [x] Squash commits?

I've tested the changes by running the following:

```
PACKER_VAR_FILES=/tmp/packer-ova.json FLATCAR_VERSION=3033.2.4 SSH_PRIVATE_KEY_FILE=/tmp/packer-key SSH_PUBLIC_KEY_FILE=/tmp/packer-key.pub make build-node-ova-vsphere-flatcar
```

Sample `/tmp/packer-ova.json` contents:

```JSON
{
    "vcenter_server":"vsphere.example.com",
    "username":"foo",
    "password":"bar",
    "datacenter": "MyDC",
    "datastore":"My Datastore",
    "folder": "foo",
    "cluster": "10.1.2.3",
    "network": "VM Network",
    "kubernetes_semver": "v1.20.10",
    "convert_to_template": "true"
}
```

johananl commented:
/test all

johananl commented:
/retest

johananl commented:
/test pull-ova-all

johananl commented:
/test pull-azure-sigs

johananl commented:
/test pull-azure-sigs

johananl commented:
/test pull-azure-sigs

johananl commented:
I've verified that the changes aren't breaking the AWS and Azure image build processes.

johananl commented:
I've broken out the generic (non-vSphere) commits into https://github.com/kubernetes-sigs/image-builder/pull/827 and https://github.com/kubernetes-sigs/image-builder/pull/829.

johananl commented:
#827 is merged.
/unhold

codenrhoden commented on images/capi/hack/image-build-ova.py:
```diff
@@ -118,6 +118,8 @@ def main():
                  "rhel7-64": {"id": "80", "version": "7", "type": "rhel7_64guest"},
                  "ubuntu-64": {"id": "94", "version": "", "type": "ubuntu64Guest"},
+                 "coreos-64": {"id": "101", "version": "", "type": "linux-64"},
+                 "flatcar-64": {"id": "101", "version": "", "type": "linux-64"},
```
What does the flatcar entry here get referenced by? I see the coreos reference.

johananl replied:
Good question. IIUC the values in this data structure are vSphere-specific, undocumented values used as OVA metadata. My guess is that @dongsupark used the CoreOS values since Flatcar is derived from CoreOS.

AFAIK these metadata are used by vSphere to indicate things like the OVA's distro to the user when deploying VMs (but you might know more than I do on this topic @codenrhoden 🙂).

kkeshavamurthy replied:
`flatcar.json` -> `packer-manifest` -> here -> OVF

johananl replied:
@codenrhoden @kkeshavamurthy do you folks have a change suggestion here? I'm happy to use another value for the Flatcar entry if anything makes more sense than the current state.

codenrhoden reviewed (commented):
Thanks for working on getting this added, @johananl! It's always a lot of work adding a new OS, and Flatcar has some unique challenges to it. :)

I've got an initial round of feedback for now. I'll try to take this for a spin someone soon and try out building an OVA for myself (sorry, I haven't tried yet).

johananl commented:
I've encountered a race which may cause Flatcar to update during image creation - see https://github.com/kubernetes-sigs/image-builder/issues/854. This bug must be fixed before we can reliably create OVA Flatcar images (and likely QEMU, too).

/hold

johananl commented on images/capi/packer/ova/centos-7.json:
```diff
@@ -1,4 +1,5 @@
 {
+  "ansible_os_family": "RedHat",
```
I think you're right, I'm not sure why we added these manually. Let me double check and correct as necessary.

johananl replied:
I've removed all of these variables as they are indeed useless. Thanks for spotting this!

johananl commented on images/capi/packer/ova/flatcar.json:
```diff
@@ -0,0 +1,25 @@
+{
+  "ansible_extra_vars": "guestinfo_datasource_slug={{user `guestinfo_datasource_slug`}} guestinfo_datasource_ref={{user `guestinfo_datasource_ref`}} guestinfo_datasource_script={{user `guestinfo_datasource_script`}} ansible_python_interpreter=/opt/bin/python",
+  "ansible_os_family": "Flatcar",
```
Yes, I agree. Will fix.

johananl replied:
Fixed!

kkeshavamurthy commented on images/capi/packer/ova/packer-node.json:
```diff
@@ -323,6 +323,14 @@
     }
   ],
   "provisioners": [
+    {
+      "environment_vars": [
+        "BUILD_NAME={{user `build_name`}}"
+      ],
+      "execute_command": "sudo -S -E {{.Vars}} bash '{{.Path}}'",
+      "script": "./packer/files/bootstrap-flatcar.sh",
+      "type": "shell"
+    },
```
Seems like this will be run on all builds? Maybe have an `if` condition in `inline`?

johananl replied:
Actually we're already doing just that for QEMU:

https://github.com/kubernetes-sigs/image-builder/blob/e10d50438ad71ad86e3967b9a72b952328792acc/images/capi/packer/qemu/packer.json#L52

Let me open a separate PR to align this provisioner across all providers. I'll then rebase this PR.

kkeshavamurthy replied:
Great. Thank you.

johananl replied:
@kkeshavamurthy done: https://github.com/kubernetes-sigs/image-builder/pull/883

johananl commented on images/capi/Makefile:
```diff
@@ -534,6 +536,7 @@ build-node-ova-local-all: $(NODE_OVA_LOCAL_BUILD_TARGETS) ## Builds all Node OVA
 build-haproxy-ova-local-photon-3: ## Builds Photon 3 HAProxy OVA w local hypervisor
 
 build-node-ova-vsphere-centos-7: ## Builds CentOS 7 Node OVA and template on vSphere
+build-node-ova-vsphere-flatcar: ## Builds Flatcar stable Node OVA and template on vSphere
```
I'll resolve this. Please feel free to reopen if anything else is missing.

johananl commented:
/test pull-azure-sigs

johananl commented:
> I've encountered a race which may cause Flatcar to update during image creation - see #854. This bug must be fixed before we can reliably create OVA Flatcar images (and likely QEMU, too).
> 
> /hold

Inadvertent updates for OVA images are handled by #895.
/unhold

johananl commented:
/test pull-azure-sigs

johananl commented:
/retest

johananl commented:
/test pull-azure-sigs

johananl commented:
/test pull-azure-sigs

johananl commented:
/test pull-azure-sigs

johananl commented:
/test pull-azure-sigs

johananl commented:
Should `build-node-ova-local-*` be a blocker for OVA changes? I'm having a hard time getting a working VMware environment to test these targets. Other than that we're good for merging as far as I'm concerned.

kkeshavamurthy commented on images/capi/packer/ova/flatcar.json:
```diff
@@ -0,0 +1,25 @@
+  "systemd_prefix": "/etc/systemd",
+  "sysusr_prefix": "/opt",
+  "sysusrlocal_prefix": "/opt",
+  "vsphere_guest_os_type": "coreos64Guest"
+}
```
Correct me if I'm wrong, 
Based on this https://github.com/flatcar-linux/Flatcar/issues/154, 
I think this should be `other3xLinux64Guest` and id which goes in `image-builder-ova.py` should be `100` (https://github.com/flatcar-linux/scripts/pull/78/files)

johananl replied:
I feel like your guess is as good as mine here @kkeshavamurthy 🙂 🤷 Let me change the two values and see what gets built.

johananl replied:
I am still very unclear about what these metadata actually influence, however I've done the following:

- Changed `vsphere_guest_os_type` from `coreos64Guest` to `other3xLinux64Guest`.
- Changed the `id` field of the `flatcar-64` element of the `OS_id_map` dict in `image-build-ova.py` from `101` to `100`.
- Changed `guest_os_type` from `coreos-64` to `flatcar-64`. This seems to be necessary since `image-build-ova.py` uses `guest_os_type` to perform a lookup in `OS_id_map`.

Following the changes above, I got a working build, hopefully with the right metadata in the resulting OVA (I have no idea how to verify that - please enlighten me in case you do).

johananl commented:
/retest

johananl commented:
/test pull-ova-all

johananl commented:
/test pull-ova-all

kkeshavamurthy commented on images/capi/hack/image-build-ova.py:
```diff
@@ -112,6 +112,8 @@ def main():
                  "centos8-64": {"id": "107", "version": "8", "type": "centos8-64"},
                  "rhel7-64": {"id": "80", "version": "7", "type": "rhel7_64guest"},
                  "ubuntu-64": {"id": "94", "version": "", "type": "ubuntu64Guest"},
+                 "coreos-64": {"id": "101", "version": "", "type": "linux-64"},
```
I think `coreos-64` is not needed anymore since `flatcar-64` is being piped in from `flatcar.json`

johananl replied:
Good point. Fixed.

kkeshavamurthy commented:
@johananl left a comment. other than that, lgtm. 
Thanks a lot for working on this PR.

johananl commented:
> @johananl left a comment. other than that, lgtm. Thanks a lot for working on this PR.

Thanks a lot @kkeshavamurthy. I've addressed your comment. If there are no further concerns I'll squash the commits and get a green CI so that we can merge.

johananl commented:
/test pull-azure-sigs

johananl commented:
/retest

johananl commented:
CI is ✅

kkeshavamurthy commented:
Great. Can you please squash the commits

johananl commented:
/test pull-azure-sigs

johananl commented:
/test pull-azure-sigs

johananl commented:
/test pull-azure-sigs

johananl commented:
/test pull-azure-sigs

johananl commented:
/test pull-azure-sigs

johananl commented:
> Great. Can you please squash the commits

Done @kkeshavamurthy.

kkeshavamurthy commented:
/lgtm
/approve
